### PR TITLE
[WIP] surrogateescape

### PR DESCRIPTION
**Pull request by slide**

### CODE REVIEW

**Comment by slide** [comment]
Close/reopen for build

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 1]
```diff
@@ -1725,11 +1725,103 @@ public static class StringOps {
...
+                private int _index = -1;
+                private char[] _output = new char[4];
+
+                public SurrogateEscapeDecoderFallbackBuffer(SurrogateEscapeDecoderFallback fallback) {
+                    _fallback = fallback;
+                }
+
+                public override bool Fallback(byte[] bytesUnknown, int index) {
+                    if (_count > 1) {
+                        ThrowLastBytesRecursive(bytesUnknown);
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: Do we really need a recursion check? This is purely an internal class.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 2]
```diff
@@ -1725,11 +1725,103 @@ public static class StringOps {
...
+                    }
+
+                    _count = 0;
+                    while (_count < 4 && _count < bytesUnknown.Length) {
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: Shouldn't we be offsetting by index or something?

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 3]
```diff
@@ -1725,11 +1725,103 @@ public static class StringOps {
...
+        
+        private class SurrogateEscapeDecoderFallback : DecoderFallback {
+            private class SurrogateEscapeDecoderFallbackBuffer : DecoderFallbackBuffer {
+                private SurrogateEscapeDecoderFallback _fallback;
+                private int _count = -1;
+                private int _index = -1;
+                private char[] _output = new char[4];
+
+                public SurrogateEscapeDecoderFallbackBuffer(SurrogateEscapeDecoderFallback fallback) {
+                    _fallback = fallback;
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: Any reason we need to save the decoder fallback? Doesn't seem to be used.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 4]
```diff
@@ -1725,11 +1725,103 @@ public static class StringOps {
 
 #if FEATURE_ENCODING
         private static DecoderFallback ReplacementFallback = new DecoderReplacementFallback("\ufffd");
+        
+        private class SurrogateEscapeDecoderFallback : DecoderFallback {
+            private class SurrogateEscapeDecoderFallbackBuffer : DecoderFallbackBuffer {
+                private SurrogateEscapeDecoderFallback _fallback;
+                private int _count = -1;
+                private int _index = -1;
+                private char[] _output = new char[4];
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: Maybe I just don't understand how these work. Why is our buffer only 4? Will need to look into how these work some more.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 5]
```diff
@@ -1821,8 +1914,66 @@ public static class StringOps {
...
+#if FEATURE_ENCODING
+        private class SurrogateEscapeEncoderFallback : EncoderFallback {
+            private class SurrogateEscapeEncoderFallbackBuffer : EncoderFallbackBuffer {
+                private SurrogateEscapeEncoderFallback _fallback;
+                private int _count = -1;
+                private int _size;
+                private char _value;
+
+                public SurrogateEscapeEncoderFallbackBuffer(SurrogateEscapeEncoderFallback fallback) {
+                    _fallback = fallback;
```
Comment (slozier) [inline Src/IronPython/Runtime/Operations/StringOps.cs]: Why save the fallback?

**Pull request closed without merging**